src/main/kotlin/no/nav/tilleggsstonader/sak/vedtak/OpphørValideringService.kt

```diff
@@ -0,0 +1,81 @@
+package no.nav.tilleggsstonader.sak.vedtak
+
+import no.nav.tilleggsstonader.sak.behandling.domain.Saksbehandling
+import no.nav.tilleggsstonader.sak.infrastruktur.exception.brukerfeilHvis
+import no.nav.tilleggsstonader.sak.vedtak.barnetilsyn.domain.BeregningsresultatTilsynBarn
+import no.nav.tilleggsstonader.sak.vilkår.stønadsvilkår.VilkårService
+import no.nav.tilleggsstonader.sak.vilkår.stønadsvilkår.domain.Vilkår
+import no.nav.tilleggsstonader.sak.vilkår.stønadsvilkår.domain.VilkårStatus
+import no.nav.tilleggsstonader.sak.vilkår.stønadsvilkår.domain.Vilkårsresultat
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.VilkårperiodeService
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.domain.ResultatVilkårperiode
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.domain.Vilkårperioder
+import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.felles.Vilkårstatus
+import org.springframework.stereotype.Service
+import java.time.LocalDate
+
+@Service
+class OpphørValideringService(
+    private val vilkårsperiodeService: VilkårperiodeService,
+    private val vilkårService: VilkårService,
+) {
+
+    fun validerPerioder(saksbehandling: Saksbehandling) {
+        val vilkår = vilkårService.hentVilkår(saksbehandling.id)
+        val vilkårperioder = vilkårsperiodeService.hentVilkårperioder(saksbehandling.id)
+
+        validerIngenNyeOppfylteVilkårEllerVilkårperioder(vilkår, vilkårperioder)
+        validerIngenEndredePerioderMedTomEtterOpphørsdato(
+            vilkårperioder,
+            vilkår,
+            saksbehandling.revurderFra ?: error("Revurder fra er påkrevd for opphør"),
+        )
+    }
+
+    fun validerIngenUtbetalingEtterOpphør(
+        beregningsresultatTilsynBarn: BeregningsresultatTilsynBarn,
+        opphørsDato: LocalDate?,
+    ) {
+        brukerfeilHvis(opphørsDato == null) { "Revurder fra dato er påkrevd for opphør" }
+        beregningsresultatTilsynBarn.perioder.forEach { periode ->
+            periode.beløpsperioder.forEach {
+                brukerfeilHvis(it.dato >= opphørsDato) { "Det er utbetalinger etter opphørsdato" }
+            }
+        }
+    }
+
+    private fun validerIngenNyeOppfylteVilkårEllerVilkårperioder(
+        vilkår: List<Vilkår>,
+        vilkårperioder: Vilkårperioder,
+    ) {
+        val finnesOppfyltVilkårEllerVilkårperioderMedStatusNy =
+            vilkår.any { it.status == VilkårStatus.NY && it.resultat == Vilkårsresultat.OPPFYLT } ||
+                vilkårperioder.målgrupper.any { it.status == Vilkårstatus.NY && it.resultat == ResultatVilkårperiode.OPPFYLT } ||
+                vilkårperioder.aktiviteter.any { it.status == Vilkårstatus.NY && it.resultat == ResultatVilkårperiode.OPPFYLT }
+
+        brukerfeilHvis(finnesOppfyltVilkårEllerVilkårperioderMedStatusNy) { "Det er nye inngangsvilkår eller vilkår som er oppfylt." }
```

Comment:
Kan dette splittes opp evt? Så får de litt mer pekepinn på hvilken side det er snakk om? Tror heller ikke de har et forhold til "vilkår", så her må vi kanskje bruke et annet ord

Reply:
Ryddig, vi prøver oss på utgifter i stedet for vilkår? [396d3cb](https://github.com/navikt/tilleggsstonader-sak/pull/483/commits/396d3cbb75c7f13b4af4434233a419dd15438748)